internal/credential/vault/repository_ssh_certificate_credential_library.go

```diff
@@ -0,0 +1,341 @@
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("in credential store: %s", l.StoreId)))
+	}
+	return newSSHCertificateCredentialLibrary, nil
+}
+
+// UpdateSSHCertificateCredentialLibrary updates the repository entry for l.PublicId with
+// the values in l for the fields listed in fieldMaskPaths. It returns a
+// new SSHCertificateCredentialLibrary containing the updated values and a count of the
+// number of records updated. l is not changed.
+//
+// l must contain a valid PublicId. Name, Description, VaultPath, Username,
+// KeyType, KeyBits, Ttl, KeyId, CriticalOptions, and Extensions can be updated. If
+// l.Name is set to a non-empty string, it must be unique within l.StoreId.
+//
+// An attribute of l will be set to NULL in the database if the attribute
+// in l is the zero value and it is included in fieldMaskPaths.
+func (r *Repository) UpdateSSHCertificateCredentialLibrary(ctx context.Context, projectId string, l *SSHCertificateCredentialLibrary, version uint32, fieldMaskPaths []string, _ ...Option) (*SSHCertificateCredentialLibrary, int, error) {
+	const op = "vault.(Repository).UpdateSSHCertificateCredentialLibrary"
+	if l == nil {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing SSHCertificateCredentialLibrary")
+	}
+	if l.SSHCertificateCredentialLibrary == nil {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing embedded SSHCertificateCredentialLibrary")
+	}
+	if l.PublicId == "" {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidPublicId, op, "missing public id")
+	}
+	if version == 0 {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing version")
+	}
+	if projectId == "" {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing project id")
+	}
+	l = l.clone()
+
+	for _, f := range fieldMaskPaths {
+		switch {
+		case strings.EqualFold(nameField, f):
+		case strings.EqualFold(descriptionField, f):
+		case strings.EqualFold(vaultPathField, f):
+		case strings.EqualFold(usernameField, f):
+		case strings.EqualFold(keyTypeField, f):
+		case strings.EqualFold(keyBitsField, f):
+		case strings.EqualFold(ttlField, f):
+		case strings.EqualFold(keyIdField, f):
+		case strings.EqualFold(criticalOptionsField, f):
+		case strings.EqualFold(extensionsField, f):
+
+		default:
+			return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidFieldMask, op, f)
+		}
+	}
+	var dbMask, nullFields []string
+	dbMask, nullFields = dbw.BuildUpdatePaths(
+		map[string]any{
+			nameField:            l.Name,
+			descriptionField:     l.Description,
+			vaultPathField:       l.VaultPath,
+			usernameField:        l.Username,
+			keyTypeField:         l.KeyType,
+			keyBitsField:         l.KeyBits,
+			ttlField:             l.Ttl,
+			keyIdField:           l.KeyId,
+			criticalOptionsField: l.CriticalOptions,
+			extensionsField:      l.Extensions,
+		},
+		fieldMaskPaths,
+		nil,
```

Comment:
You need to add `keyBitsField` here so it is allowed to be `0`. Othwerwise `dbw.BuildUpdatePaths` will convert the `0` to a `null`, which is what is causing the null constraint error when trying to update KeyBits to 0.

```suggestion
[]string{keyBitsField},
```

Reply:
Thanks. I'm sad I didn't pursue this. I saw the other credential library didn't use this field and I assumed this was not the problem